tool/src/dart/class.rs

```diff
@@ -0,0 +1,914 @@
+
+        if let Some(param_self) = method.param_self.as_ref() {
+            param_types_ffi.push(self.gen_self_type_ffi(&param_self.ty, false));
+            param_types_ffi_cast.push(self.gen_self_type_ffi(&param_self.ty, true));
+            dart_to_ffi_params.push(self.gen_dart_to_c_self(&param_self.ty));
+        }
+
+        let mut slice_conversions = Vec::new();
+
+        for param in method.params.iter() {
+            param_decls_dart.push(self.gen_ty_decl(&param.ty, param.name.as_str()));
+            let param_type_ffi = self.gen_type_name_ffi(&param.ty, false);
+            if let Some(element_type_ffi) = param_type_ffi.strip_prefix("Slice:") {
+                param_types_ffi.push(format!("ffi.Pointer<{element_type_ffi}>").into());
+                param_types_ffi.push("ffi.Size".into());
+            } else {
+                param_types_ffi.push(param_type_ffi);
+            }
+            let param_type_ffi_cast = self.gen_type_name_ffi(&param.ty, true);
+            if let Some(element_type_ffi_cast) = param_type_ffi_cast.strip_prefix("Slice:") {
+                param_types_ffi_cast.push(format!("ffi.Pointer<{element_type_ffi_cast}>").into());
+                param_types_ffi_cast.push("int".into());
+            } else {
+                param_types_ffi_cast.push(param_type_ffi_cast);
+            }
+            let conversions = self.gen_dart_to_c_for_type(
+                &param.ty,
+                self.cx.formatter.fmt_param_name(param.name.as_str()),
+                &mut slice_conversions,
+            );
+            dart_to_ffi_params.extend(
+                conversions
+                    .into_iter()
+                    .map(|PartiallyNamedExpression { expression, .. }| expression),
+            );
+        }
+
+        if method.is_writeable() {
+            dart_to_ffi_params.push("writeable._underlying".into());
+            param_types_ffi.push("ffi.Pointer<ffi.Opaque>".into());
+            param_types_ffi_cast.push("ffi.Pointer<ffi.Opaque>".into());
+            self.helper_classes
+                .insert("writeable".into(), include_str!("writeable.dart").into());
+        }
+
+        let ffi_return_ty = self.gen_ffi_return_type_name(&method.output, false);
+        let ffi_cast_return_ty = self.gen_ffi_return_type_name(&method.output, true);
+
+        let dart_return_expression: Option<Cow<str>> =
+            self.gen_c_to_dart_for_return_type(&method.output, "result".into());
+
+        let params = param_decls_dart
+            .iter()
+            .map(|param| format!("{} {}", param.type_name, param.var_name))
+            .collect::<Vec<_>>()
+            .join(", ");
+
+        let return_ty = self.gen_dart_return_type_name(&method.output);
+        let method_name = self.cx.formatter.fmt_method_name(method);
+        let declaration = if method.param_self.is_none() {
+            if return_ty == type_name {
+                format!("factory {type_name}.{method_name}({params})")
```

Comment:
thought: this is the kind of logic Shane likes to have in the template but I'm not so sure

Reply:
This is a single line and too complex to handle in the template